spec.hpp

```diff
@@ -12,6 +12,11 @@ template <typename Err>
     throw Err(msg);
 }
 
+template <typename Err>
+[[noreturn]] static void __attribute__((preserve_most)) error(std::string msg) {
+    throw Err(msg);
```

Comment:
```suggestion
    // todo: check if forwarding args to `std::string` allows for nicer error messages in hot path errors (or just figure out other system to allow dynamic-ness
    throw Err(msg);
```
butchered some of the error messages when I was trying to speed up validation, def a good area for improvement